File: LeetcodePractice/src/main/java/com/Bernie/leetcode/editor/cn/[面试题 04.02]最小高度树.java
//给定一个有序整数数组，元素各不相同且按升序排列，编写一个算法，创建一棵高度最小的二叉搜索树。示例: 给定有序数组: [-10,-3,0,5,9], 一个可能
//的答案是：[0,-3,9,-10,null,5]，它可以表示下面这个高度平衡二叉搜索树：           0          / \        -3 
//  9        /   /      -10  5 Related Topics 树 深度优先搜索 
// 👍 86 👎 0


package com.Bernie.leetcode.editor.cn;

class MinimumHeightTreeLcci {
    public static void main(String[] args) {
        Solution solution = new MinimumHeightTreeLcci().new Solution();
    }
    //leetcode submit region begin(Prohibit modification and deletion)

    /**
     * Definition for a binary tree node.*/
      public class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode(int x) { val = x; }
      }

    class Solution {
        public TreeNode sortedArrayToBST(int[] nums) {
            return build(nums, 0, nums.length -1);
        }

        private TreeNode build(int[] nums, int left, int right) {
            if (left > right) {
                return null;
            }

            int mid = (right -left)/2 + left;
            TreeNode root = new TreeNode(nums[mid]);
            root.left = build(nums, left, mid -1);
            root.right = build(nums, mid+1, right);
            return root;
        }
    }
//leetcode submit region end(Prohibit modification and deletion)

}

File: LeetcodePractice/src/main/java/com/Bernie/leetcode/editor/cn/[面试题 04.02]最小高度树.py
# 给定一个有序整数数组，元素各不相同且按升序排列，编写一个算法，创建一棵高度最小的二叉搜索树。示例: 给定有序数组: [-10,-3,0,5,9], 一个可能
# 的答案是：[0,-3,9,-10,null,5]，它可以表示下面这个高度平衡二叉搜索树：           0          / \        -3 
#   9        /   /      -10  5 Related Topics 树 深度优先搜索 
#  👍 86 👎 0


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
from typing import List


class TreeNode:
     def __init__(self, x):
         self.val = x
         self.left = None
         self.right = None



class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        if not nums:
            return
        mid = len(nums) // 2
        root = TreeNode(nums[mid])
        root.left = self.sortedArrayToBST(nums[ : mid])
        root.right = self.sortedArrayToBST(nums[mid+1: ])
        return root

solution = Solution()
print(solution.sortedArrayToBST([-10,-3,0,5,9]))
# leetcode submit region end(Prohibit modification and deletion)
